packages/lib/src/vis/shared/SelectionTool.tsx

```diff
@@ -26,78 +28,96 @@ function SelectionTool(props: Props) {
     modifierKey,
   } = props;
 
-  const [startPoint, setStartPoint] = useState<Vector2>();
-  const [endPoint, setEndPoint] = useState<Vector2>();
-  const [isAllowed, toggleAllowed] = useToggle(!modifierKey);
   const camera = useThree((state) => state.camera);
-
   const { worldToData } = useAxisSystemContext();
+
+  const [startPoint, setStartPoint] = useState<Vector2>();
+  const [endPoint, setEndPoint] = useRafState<Vector2 | undefined>(undefined);
+  const [isVisible, toggleVisible] = useToggle(true);
+
   const onPointerDown = useCallback(
     (evt: CanvasEvent<PointerEvent>) => {
-      if (!isAllowed) {
+      const { unprojectedPoint, sourceEvent } = evt;
+      if (!checkModifierKey(modifierKey, sourceEvent)) {
         return;
       }
 
-      const { unprojectedPoint, sourceEvent } = evt;
       const { target, pointerId } = sourceEvent;
       (target as Element).setPointerCapture(pointerId);
-
       setStartPoint(worldToData(unprojectedPoint));
-      setEndPoint(undefined);
-      onSelectionStart?.();
+
+      if (onSelectionStart) {
+        onSelectionStart();
+      }
```

Comment:
Reverting to a clearer syntax, since I'm going over the complexity limit anyway...